spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DatastoreServiceObjectToKeyFactory.java

```diff
@@ -91,17 +90,34 @@ public Key getKeyFromObject(Object entity,
 
 	@Override
 	public Key allocateKeyForObject(Object entity,
-			DatastorePersistentEntity datastorePersistentEntity) {
+			DatastorePersistentEntity datastorePersistentEntity, Key... ancestors) {
 		Assert.notNull(entity, "Cannot get key for null entity object.");
 		Assert.notNull(datastorePersistentEntity, "Persistent entity must not be null.");
 		PersistentProperty idProp = datastorePersistentEntity.getIdPropertyOrFail();
-		Key allocatedKey = this.datastore.allocateId(
-				getKeyFactory().setKind(datastorePersistentEntity.kindName()).newKey());
+
+		KeyFactory keyFactory = getKeyFactory().setKind(datastorePersistentEntity.kindName());
+		if (ancestors != null) {
+			for (Key ancestor : ancestors) {
+				keyFactory.addAncestor(ancestor.getName() != null
+						? PathElement.of(ancestor.getKind(), ancestor.getName())
+						: PathElement.of(ancestor.getKind(), ancestor.getId()));
+			}
+		}
+		Key allocatedKey = this.datastore.allocateId(keyFactory.newKey());
 		Class idPropType = idProp.getType();
-		Object id = idPropType.equals(Long.class) ? allocatedKey.getId() : allocatedKey.getId().toString();
 
-		datastorePersistentEntity.getPropertyAccessor(entity).setProperty(idProp,
-				idPropType.equals(Key.class) ? allocatedKey : id);
+		Object value;
+		if (idPropType.equals(Key.class)) {
+			value = allocatedKey;
+		}
+		else if (idPropType.equals(Long.class)) {
+			value = allocatedKey.getId();
```

Comment:
Check if updating without ancestor info causes a failure on Datastore.

Reply:
@meltsufin @dmitry-s I just did this test manually. The parent is indeed part of the key, so you can have two entities with the same ID value but with different parents since those are 2 separate keys.